### PR TITLE
Test and document environment inside `container` step with current shortcomings

### PR DESCRIPTION
**Pull request by Vlatombe**
### Testing done

```[tasklist]
### Submitter checklist
- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecoratorPipelineTest.java [hunk 1]
```diff
@@ -115,4 +115,21 @@ public void envVarDollarSignEscaping() throws Exception {
...
+                "from Groovy: /opt/java/openjdk/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin", b);
+        r.assertLogContains(
+                "outside container: /opt/java/openjdk/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
+                b);
+        r.assertLogContains(
+                "outside container with override: /bar:/opt/java/openjdk/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
+                b);
+        r.assertLogContains("inside container: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin", b);
+        r.assertLogContains(
+                "inside container with override: /bar:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin", b);
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecoratorPipelineTest.java:133]: FTR currently fails with

```
inside container with override: /bar:/opt/java/openjdk/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecoratorPipelineTest.java [hunk 2]
```diff
@@ -115,4 +115,21 @@ public void envVarDollarSignEscaping() throws Exception {
         r.assertLogContains("outside container: $string$with$dollars", b);
         r.assertLogContains("inside container: $string$with$dollars", b);
     }
+
+    @Test
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecoratorPipelineTest.java:119]: Could slap an `@Ignore` on it and make this mergeable as a reminder of what is broken.

File: src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/containerEnvironmentIsHonored.groovy [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  - name: alpine
+    image: alpine:3.19.0
+    # PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+    command:
+    - sleep
+    args:
+    - infinity
+''') {
+  node(POD_LABEL) {
+    echo "from Groovy: ${env.PATH}"
```
Comment (jglick) [inline src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/containerEnvironmentIsHonored.groovy:16]: Also interesting to check this at various points. Your proposed fix using Bourne shell expansion (`eval`?) in the `ContainerExecDecorator` might fix the `inside container with override` from `sh` but would not fix it from Groovy. Whether anyone _cares_ is another question, it just might be a source of confusion if the Groovy view of the environment differs from what `sh` steps actually use.

**Comment by jglick** [comment]
I suspect the problem is that some code such as https://github.com/jenkinsci/kubernetes-plugin/blob/6ab2953c345cc0ffa16240af920f14537fe24bcf/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java#L338-L348 does not belong in this class at all, and should instead be in https://github.com/jenkinsci/kubernetes-plugin/blob/6ab2953c345cc0ffa16240af920f14537fe24bcf/src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerStepExecution.java#L51-L53 https://github.com/jenkinsci/kubernetes-plugin/pull/1486#discussion_r1432892359 might make this more apparent.

File: src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/containerEnvironmentIsHonored.groovy (hunk 3 above)
Comment (Vlatombe) [reply src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/containerEnvironmentIsHonored.groovy:16]: Pure groovy ought to be always executed on the controller, independently from whether we are within a node or a container, so to be this is not a problem.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecoratorPipelineTest.java [hunk 4]
```diff
@@ -121,15 +121,32 @@ public void containerEnvironmentIsHonored() throws Exception {
...
         r.assertLogContains(
-                "inside container with override: /bar:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin", b);
+                "from Groovy inside container: /opt/java/openjdk/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
+                b);
+        r.assertLogContains(
+                "from shell inside container: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin", b);
+        // When using groovy, the environment relies on the computer's environment, using container step doesn't change
+        // anything.
+        r.assertLogContains(
+                "from Groovy inside container with override: /bar:/opt/java/openjdk/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
```
Comment (jglick) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecoratorPipelineTest.java:144]: My point in https://github.com/jenkinsci/kubernetes-plugin/pull/1486#discussion_r1432892359 & https://github.com/jenkinsci/kubernetes-plugin/pull/1486#issuecomment-1864729718 is that here we should actually expect `/bar:/usr/local/sbin:…` just like from the `sh` step, even if we do not consider it _important_ to fix.

**Review by jglick — APPROVED** [review]

**Pull request merged**